### ISSUE TITLE
[BUG] PPS Errors with BSB-LAN Versions > 2.0.16

### ISSUE THREAD

**Issue by malansk** [fredlcore/BSB-LAN#615]
**BSB-LAN Version**
2.0.16 (ok but hangs after 3months uptime)
2.0.113 (PPS "Betriebsart" not switchable, PPS "Zieltemperatur" always 0°C -> Heater not running)
2.2.1 (PPS-Error Displayed on Heater Control Module)
3.x (PPS-Error Displayed on Heater Control Module) 

**Architecture**
Arduino Due w. LAN shield

**Bus system**
PPS (Siemens RVA53.242 from year approx. 2008)

**Describe the bug**
I have Version 2.0.16 running since 12/2020 but after approx. 3 months uptime I always need to restart the Arduino Due because BSB-LAN is no longer running (Web-Interface not reachable, WGET from Smarthome system gives no values)

Today I tried newer Versions of BSB-LAN but they are missing some PPS values or they block the PPS bus on the heater control module if PPS write is enabled.

**Version 2.0.113**
PPS "Betriebsart" can be changed but no change on heater control module 
PPS "Zieltemperatur" always shows 0°C (with Version 2.0.16 it shows 15°C or 24°C depending on Presence Switch)
Heater is not running in Automatic mode (only OFF mode w. Freeze-Protection)

**Version 2.2.1 & 3.x**
As soon as the Arduino Due is up and running, I get "Err" Message on the Heater Control Module. It shows Error 146 - PPS-Bus-Error
If I change the configuration from "Raumregelgerät" to "Passive" (no writing on PPS bus) the Error dissappears and I get some PPS values but not all the PPS values I got with 2.0.16

I am uploading each version from different folder on hard-drive, with the config files renamed (from .default) and with the necessary configuration settings (e.g. Bus Type "PPS" etc.).
I have even cleared the EPROM content by command line switch /NE

After Downgrading to 2.0.16, everything works ok (but after 3 months uptime needs restart).

**To Reproduce**

**Logs**
If possible, attach or copy/paste a log of the Arduino IDE's serial monitor when performing the above actions. If the error is reproducible, do so after pressing the reset button on the board.

**Expected behavior**
Versions 2.0.113 and 2.2.1 should give all of the PPS values from Version 2.0.16.
I have read that the category numbers for PPS values are changed, this is not the problem.
PPS bus should not be blocked by Version 2.2.1 or higher if PPS write on bus is enabled.

Can I use the PPS category definiton of version 2.0.16 (in BSB_LAN_defs.h) with the newer versions ?
Or can I somehow disable some PPS command / change their physical address/identifiert to make them work again like in 2.0.16 ?

//-> upon further investigation of the BSB_lan_defs.h and BSB_lan files, the PPS bus commands in BSB_LAN_defs.h did not change (only their category number).
//-> but the heater bus handling (send/receive date) in BSB_lan.io sketch looks very different in the new versions if I compare it to the 2.0.16 version from 2020. So maybe the new heater bus code routines are not working propperly with my controller (even though the both should emulate QAA70 device).

**Screenshots**

**Desktop (if applicable, please complete the following information):**
 - OS: Windows 10
 - Browser Chrome 
 - Browser version 116.0.5845.141 (Offizieller Build) (32-Bit)

**Additional context**

**Attach your BSB_LAN_config.h file to your bug report (remove any confidential information if necessary). Bug reports without this file attached will be closed immediately.**
[BSB_lan_config_V2.0.16.zip](https://github.com/fredlcore/BSB-LAN/files/13999309/BSB_lan_config_V2.0.16.zip)
[BSB_LAN_config_V2.0.113.zip](https://github.com/fredlcore/BSB-LAN/files/13999311/BSB_LAN_config_V2.0.113.zip)
[BSB_LAN_config_V2.2.1.zip](https://github.com/fredlcore/BSB-LAN/files/13999315/BSB_LAN_config_V2.2.1.zip)

**Comment by fredlcore** [comment]
Since I no longer have access to a PPS controller, I would need serial monitor log files from all three versions, both when running in passive mode as well as in room unit mode. The logs need to begin when booting the Arduino and need to run for 1-2 minutes and during that time you should change the problematic parameters.

**Comment by malansk** [comment]
Yes, I was trying to get serial monitor log files from the Arduino Due but was not successfull. Later I found some reference posts that one the Due it‘s not possible to get the SerialMonitor output from the same port like the programming was done, so I will try to open the case and switch over the USB cable from native port to programming port to get some logfile. Would logfile from SD-Card also work ? I saw some options to enable logging of telegrams.

**Comment by fredlcore** [comment]
No, that is wrong information. Yes, the Due has two USB ports, but for BSB-LAN, both programming as well as serial output is via the programming port. And no, logging to SD card is not enough in case of troubleshooting as the logs do not cover both sending and receiving data.

**Comment by fredlcore** [comment]
I have had reports from PPS users using the most recent BSB-LAN version on a rather old (even for PPS) RVA36 controller. So I assume your situation is a bit more specific. With the said log files, I can nevertheless have a look, but without these, there's nothing I can do. So for the moment, I'm going to close this issue until there is further background information.

**Comment by malansk** [comment]
Today I had finally time to collect some log files with Serial Monitor running on the Arduino Due Programming port.

1. "PPS Bus Errors when write on bus is enabled" where probably caused by bad EEPROM content e.g. Heating start / end-times of >24h.
EEPROM was probably not cleared by /NE switch due to missing definition of #define RESET in config.h.
After properly clearing EEPROM, no more PPS BUs Errors shown in the heating controller display

2. Version 2.0.16 shows real data on almost all channels and can write to all channels which have SET button in WEB-GUI 
(Betriebsart, Präsenztaste, Außentemperatur, Vorlauftemperatur, Mischervorlauftemperatur, Trinkwassertemperatur, TWW Reduziertsollwert, TWW Nennsollwert, Komfortsollwert, Reduziertsollwert,  Frostschutzsolltwert, Sollwert Maximum, Raumtemperatur (set by BSB-LAN), Heizperioden, Uhrzeit, Zieltemperatur.

 Version 2.016 is missing data on these channel (not shure if this is bug or if the heating controller doesn't send them over PPS)
Brennerstatus, Außentemperatur gemischt, Vorlaufsollwert

3. Version 2.1 is showing the most Temperature data in the Serial Monitor with 0°C  and showing wrong values for Außentemperatur in the Serial Monitor. Many telegramms are missing at all (Brennerstatus, Heating Times).
   
4. Version 3.3.5 does not recognize any PPS telegrams at all in PPS Bus read mode, even though the raw messages look the same (et least the unrecognized ones). With write mode enabled, the same telegrams like in 2.1 are shown (also with 0°C for the temperatures and missing the Brennerstatus and Heating times).

6. Version 3.3.5 is not able to connect to MQTT server with PPS write mode enabled.

Please let me know if you need some more logs / information.

**Comment by malansk** [comment]
I have copied some interesting differences in the SerialMonitor_Vergleich.txt.
Is looks like some parts of the telegrams are missing the second half of the values (only 00) - the first half seems to be ok.
The values which I am able to set via WEB-GUI (e.g. Komfortsollwert or Reduziertsollwert) seems to be complete.
Maybe the PPS Bus read function truncates the second half of the telegrams ?

**Comment by malansk** [comment]
@fredlcore 
please have a look at this, I'm not shure if you see this update since the issue has already been closed.

**Comment by fredlcore** [comment]
Thanks for the intensive research, however, 3.3.5 is already several weeks/months old, current version is 3.4.6, and I also fixed a non-trivial bug for PPS in the meantime. So could you please check again with the most recent version and let me know if the problem persists? Thanks!

**Comment by malansk** [comment]
Where do I get it ?
On Github it shows 3.3 as latest...
https://github.com/fredlcore/BSB-LAN/releases

**Comment by fredlcore** [comment]
From the master repository, as described in the manual...

**Comment by malansk** [comment]
Version 3.4.6 gives me PPS Bus Error on the heating system if I set PPS bus to write - even with EEPROM cleared.
Depending on QAA Type, Serial Monitor gives "Collision on Bus detected".

Attached are the latest logfiles for read and write.

**Comment by malansk** [comment]
[SerialMonitor_3.4.6.txt](https://github.com/fredlcore/BSB-LAN/files/15221754/SerialMonitor_3.4.6.txt)
[SerialMonitor_3.4.6_write_PPSBUSERROR.txt](https://github.com/fredlcore/BSB-LAN/files/15221755/SerialMonitor_3.4.6_write_PPSBUSERROR.txt)
[SerialMonitor_Vergleich.txt](https://github.com/fredlcore/BSB-LAN/files/15221756/SerialMonitor_Vergleich.txt)

**Comment by fredlcore** [comment]
You seem to have the "monitor" function active. Don't do that, that's basically only necessary for specific debugging purposes which is why it is turned off by default. In this case, it might cause problems because PPS is a very timing sensitive protocol.
Furthermore, I  don't see bus errors in your file S.erialMonitor_3.4.6_write_PPSBUSERROR.txt.

**Comment by malansk** [comment]
I have changed #define DEFAULT_FLAG from 0 (which I was using in 2.0.16) to FL_SW_CTL_RONLY and now I am getting live PPS values from the heating controller :-)

But there are some issues:

*Mischervorlauftemperatur is shown as -512°C (this was ok in 2.0.16 but changed already in Version 2.1 to -512°C)
 (2.0.16 PPS-Bus - Mischervorlauftemperatur: 30.6 &deg;C00 2c 00 00 00 00 07 a5 00
#15004: 30.6 &deg;C
30.600000
30.600000)
(3.4.6 PPS-Bus - Mischervorlauftemperatur: -512.0 °C
00 2C 00 00 00 00 80 01 00
#15034: -512.0 °C)

*QAA Type (was "15062 PPS-Bus - QAA Modell: 83 - QAA 70fd 38 ff ff ff ff ff 53 7d" in Ver 2.0.16) is now "15046.0 PPS-Bus - QAA Modell: 65363FD 38 FF FF FF FF FF 53 7D" (Modell 83 changed to 65363 or so some other number starting with 65...).
*Betriebsart (was "15010 PPS-Bus - Betriebsart: 0 - Automatischfd 49 ff ff ff ff ff 00 bf" in Version 2.0.16 is now "PPS ANS 15000.0 PPS-Bus - Betriebsart: 65281" or some other number starting with 65...

Please check the logs for the values of QAA Type, Betriebsartumschaltung and Mischervorlauftemperatur.
By the way, "Präsenztaste" is working and I can switch it on and off.

Heating controller still shows "PPS Bus error" since Version 3.4.6 (it is not shown in the log of the serial monitor) and I can't change Betriebsart on the controller.
Maybe it is blocked by wrong QA type and wrong Betriebsart from BSB-Lan ?

I have turned off the "monitor" function off now.

**Comment by malansk** [comment]
[SerialMonitor_3.4.6_write_PPSBUSERROR_PPS_Values_read_ok_FL_SW_CTL_RONLY_MonitorOFF_Betriebsartumschaltungen.txt](https://github.com/fredlcore/BSB-LAN/files/15222997/SerialMonitor_3.4.6_write_PPSBUSERROR_PPS_Values_read_ok_FL_SW_CTL_RONLY_MonitorOFF_Betriebsartumschaltungen.txt)

**Comment by malansk** [comment]
There is no broken sensor, with 2.1.16 is shows the correct temp, with each version higher it shows -512°c.

The fix looks completely broken, for most telegrams, showing decoding error in GUI and other error in logfile (refer to log file)

**Comment by malansk** [comment]
[SerialMonitor_3.4.6_newBuilt_Broken.txt](https://github.com/fredlcore/BSB-LAN/files/15224543/SerialMonitor_3.4.6_newBuilt_Broken.txt)

**Comment by fredlcore** [comment]
Well, the imortant info (also) is that the telegrams in question before are now displayed correctly. With the bugfix I've just uploaded, it should work fine.
As for the sensor, I cannot change the telegrams coming from the heater. This is what your heater sends to BSB-LAN:
`1D 2C FF FF FF FF 80 01 3A`
The first byte (1D) indicates a broadcast information telegram from the heater to the QAA.
The second byte is indicating the type of parameter (2C = Mischervorlauftemperatur) and the 3rd- and 2nd-last bytes are the temperature (0x8001 divided by 64 (temperature operand) is -512).
In version 2.0.16 you quoted a different telegram, so that's the reason for the different temperatures displayed. As to _why_ your heater sends different telegrams with the different versions, I have no idea. Maybe you use a different QAA mode, but whatever it is, I have no PPS heating system to figure this out. If you figure out what causes the different telegrams and if this can be influenced by BSB-LAN, I can certainly try to adjust that, but otherwise, there is no chance for me to do so. 
Could it be that this specific temperature now shows up on a different parameter?

**Comment by malansk** [comment]
I have found detailed manual, it shows that the "PPS Error" I was refering to on the Heater controller is in fact Error Code "146" which is "Ungültige Anlagenkonfiguration". The Anlagenkonfiguration specifies which type of heater (1-stufig / 2-stufig / modulierend) and which type of peripherals (Pumpen, Mischer). That's why in Version newer then 2.1.16 it shows -512°C for Mischervorlauftemperatur and "Ungültige Anlagenkonfiguration" because BSB-Lan somehow changes my original "Anlagenkonfiguration" Type 15 (1-stufig, Mischerkreis) to something else if write to pps bus is enabled !
Was there an update from 2.1.16 which somehow sets heater type / family type / device type ?
In the config.h I have always used this:
===============================
/* Set the device family and device variant of your heating system. Only change this if you _really_ know what you are doing!
 * Set fixed_device_family and fixed_device_variant to your device family and variant (parameters 6225 and 6226) here
 * if heating system is not running when BSB-LAN is powered on.
*/
uint16_t fixed_device_family = 0;
uint16_t fixed_device_variant = 0;
======================
Maybe this setting was not working for PPS in 2.1.16 but was updated to actually work from 2.1.xx ?
Is there a way to disable this in the code for testing ?

**Comment by fredlcore** [comment]
If you want to try some more, you can _either_ uncomment the two `tx_msg` in lines 24 and 25 in include/pps_handling.h (case 3) _or_ in lines 28 an 29, so we can figure out which causes the issue. But still, a comparison with what your QAA sends in these cases would be helpful in order to hopefully decode what the actually mean...

**Comment by malansk** [comment]
Uncommenting case 3 or case 4 doesn't make a difference.
But uncommenting case 0 (QAA Type setting) solves the issue - no more "Ungültige Anlagenkonfiguration" Error and everything is working like in 2.1.16 !
The weird thing is that looking at the Serial monitor I get the same result (FD 38 FF FF FF FF FF 53 7D) if case 0 is active in 3.4.6 like I get from 2.1.16 so I doubt that the actual code from this line in pps_handling.h is the culprit.
But since in case 0 "pps_values[PPS_QTP] = QAA_TYPE" is set, maybe it triggers a bug later somewhere in the code.
e.g. 
"case 0x38: pps_values[PPS_QTP] = msg[7+pps_offset]; break; // QAA type"
or
"case 0x38: msg_cycle = 0; break;"

If I leave case 0 enabled and uncomment the "case 0x38: msg_cycle = 0; break;" line the heater starts and runs for approx. 5 minutes before throwing the "ungültiger Anlagentyp" error.

So it looks like there is definitly something buggy with this QQA Type / address 0x38 handling at least for my system.

**Comment by malansk** [comment]
I will let it run overnight with case 0 uncommented to see if this is truly stable,

**Comment by malansk** [comment]
[SerialMonitor_3.4.6_latestBugfix_read.txt](https://github.com/fredlcore/BSB-LAN/files/15227235/SerialMonitor_3.4.6_latestBugfix_read.txt)
[SerialMonitor_3.4.6_latestBugfix_write.txt](https://github.com/fredlcore/BSB-LAN/files/15227236/SerialMonitor_3.4.6_latestBugfix_write.txt)
[SerialMonitor_3.4.6_latestBugfix_write_case0-uncommented.txt](https://github.com/fredlcore/BSB-LAN/files/15227237/SerialMonitor_3.4.6_latestBugfix_write_case0-uncommented.txt)

**Comment by malansk** [comment]
One more thing, after restart of heating controller and BSB-LAN, I sometimes get "Verbindung Unterbrochen 1 - JA".
Maybe this is related to my uncommenting of case0 ? To get rid of that, I undid my change, restartet (which gave the error "ungültiger Anlagentyp"), did my change to uncomment case0 and everything was workinG again.

**Comment by malansk** [comment]
One more thing, "15023 PPS-Bus - TWW Nennsollwert:" is not persistend. After restart it's at 0°C and I don't get warm water. After setting it manually to 55°C in WEB-GUI, the value gets transfered to the heating controller. But after next restart of BSB-LAN, it's again at 0°C. The value should be written to EEPROM after setting and read back from EEPROM after restart.
If I cange this value in the heating controller, BSB-LAN will not register this change. Looks like writing this value to PPS is ok but reading from PPS fails.

**Comment by fredlcore** [comment]
Well, the reason the error is going away is that the heating system thinks that there is no QAA (i.e. BSB-LAN) connected and thus ignores any other kind of data from the QAA. That's why you also get the "connection interrupted" telegram with which the heating system informs any QAA to start sending telegram type 38 again. That's why it's case 0 because this has to be sent most often so that the heating system doesn't think the connection is interrupted.
So there is nothing buggy with that, but some other kind of telegram case in the message queue causes the problem. You could try to go through it one by one, but as you said that the "Anlagentyp" is not correct, it would be helpful to let me know which value it is at now that the error occurs?

For TWW Nennsollwert, I'll have to have a look, that should already be written to EEPROM, as it has FL_EEPROM flag set in BSB_LAN_defs.h. Room temperature is not written to EEPROM for safety reasons. Imagine your home automation which delivers the room temperature to BSB-LAN is no longer updating. Then the last known room temperature (say, 21.4 degrees) would continue to be written to the heater even if it is below zero already. This would at least become obvious when BSB-LAN restarts and does not get a fresh room temperature.
The other reason is that the room temperature usually changes most often and would thus lead to continous writes to the EEPROM, reducing its life-span. Since you have no other way to inform the heater about the room temperature anyway (because in write mode you have no QAA anymore that could do it for you), you have to make sure that there is a continous and reliable source of room temerature.

**Comment by fredlcore** [comment]
By the way, a room temperature of 0 sent to the heater is most likely the reason for the error, because your configuration requires a valid room temperature. 
In the unlikely case that it's not, check telegram types 4E, 55 and 56. These are the only ones that do not have a known function yet.

**Comment by malansk** [comment]
I have set a room temperatur wie WEB-GUI but it does not change the Error of "Ungültiger Anlagentyp". I have disabled Room-Temperatur correction function in the heater controller since longe time because I don't have a good room for reference anyhow.

The controller is expecting "Anlagentyp [1...150]" but shows "0" when I have write enabled. 15 would be the correct type which is shown with write disabled.

I have uncommented all 4E / 55 / 56 telegrams without change. 

In the Serial Monitor it shows "Collision on the bus, retrying".
I did not find this in any of the include files or in the .io sketch - where is this "Collision" detected & what's the retry code doing ?

**Comment by malansk** [comment]
![Collision on the bus - retrying](https://github.com/fredlcore/BSB-LAN/assets/25431062/5c6d08cf-1358-496c-9ce3-5e12ea14b1d8)

**Comment by malansk** [comment]
In pps_handlinh.h there is this line:
if ((msg[0] & 0x0F) == 0x0E) {   // Heater requests information from the QAA (i.e. BSB-LAN) with telegram type 0x1E

The value 0x0E does not seem to match the value in the comment 0x1E.
Which one is the right one ?

In V2.1.16 log I see several times:
"Unknown request: 1e 00 00 00 00 00 00 00 00"
Maybe V2.1.16 is not answering the 1E requests but 3.4.6 does and this could be the problem ?
How can I disable answering 1E requests ?

**Comment by fredlcore** [comment]
Please don't try to look for solutions in the code if you don't really understand what's going on, for example that 0x1E & 0x0F is actually 0x0E.
The only time BSB-LAN sends a zero (matching your wrong Anlagetyp) is with telegram type 0x4E. So you might want to change line 34 to 0x0F to see if that makes a difference. 
Also check what is necessary for the heater to go back to normal. Do you have to turn it off and on again or is it enough to set BSB-LAN to read-only? Maybe it takes a bit of time for the uncommented/changed telegram types to take effect?
Otherwise the only solution would be to disable each sent telegram one by one, starting from the bottom (case 26).

**Comment by malansk** [comment]
I did change line 34 to 0x0F but no difference.
The heater goes out of error mode after 5..10s after switching from write to read. I also did try several cold-start of heater and BSB-Lan.
I did uncomment each case, but the only thing that makes a difference is when I uncomment case 0 (QAA Type).
This makes me think that the case 0 is the actual problem with my system.

I made this test:
*case 0 activ (original condition)
*compile sketch, upload -> heater in Error mode, "Verbindung unterbrochen 0 - Nein"
*case 0 uncommented tx_msg's
*compile sketch, upload -> heater out of Error mode, "Verbindung unterbrochen 0 - Nein"
-> I can now use WEB-GUI to change heater mode, Presence, WW Temperatursollwert, ... so everything is working like in 2.1.6., only problem is that after next restart of BSB-LAN or heater controller is is shown "Verbindung unterbrochen 1 - JA"
Obviously my heater needs case 0 for first time initialization apower up, (but there is somewhere an error in the data leading to error message), but after initialization it is happy to accept any command from BSB-lan (except for the 38 - QAA type which throws immediate error).

**Comment by malansk** [comment]
Could you please provide a pps_handling.h stripped down to the bare minimum needed for initialization (set QAA type to QAA50, set Betriebsmode to Manual, read one known working value from heater e.g. Außentemperatur, set one known working value to heater e.g. Warmwassersollwert) without any variables & if..then..else ... switch ... case ... logic ?
Just to confirm that the low-level initialization done by QAA type announcment is working ?

**Comment by fredlcore** [comment]
Since you choose to ignore my knowledge about the PPS protocol, why should I invest hours to strip down the pps_handling.h so that it works the way you think it should work? I told you what you can do, and that is to eliminate the problematic telegram. The only telegrams that BSB-LAN sends are the `FD` telegrams, and I told you how you can disable them step by step. Maybe it's a combination of several telegrams, maybe its specific values that your heater doesn't like, who knows. If you have a QAA70, you can comare the telegrams that BSB-LAN sends with the ones the QAA sends and then figure out which telegrams need to be changed. 

Your heater will work without a `38` telegram for a few minutes, but none of the known heaters do longer than that. It is basically the heartbeat of the QAA, and since you yourself admitted, it's the same telegram that gets sent in 2.0.16.

To repeat it again: The problem stops when you disable case 0 because none of the other telegrams will be accepted by the heater, even though you keep sending them. And don't get confused by the fact that the values you change are visible in the webinterface. They will always do, even if you disconnect the heater. But check a few minutes after you disable case 0 whether these values can still be checked on the heater itself. As I said, after a few minutes (once "Verbindung unterbrochen 1 - JA" comes up), none of the values will be accepted anymore.

**Comment by fredlcore** [comment]
In your last log (SerialMonitor_3.4.6_latestBugfix_write.txt) there are a number of parameter values that can cause the error on the heater:
```
PPS ANS 15023.0 PPS-Bus - TWW Nennsollwert: 0.0 Â°C
FD 0B FF FF FF FF 00 00 FC
PPS ANS 15022.0 PPS-Bus - TWW Reduziertsollwert: 0.2 Â°C
FD 1E FF FF FF FF 00 0A DF
PPS ANS 15006.0 PPS-Bus - Frostschutzsollwert: 0.1 Â°C
FD 1B FF FF 00 00 00 06 E4
PPS ANS 15008.0 PPS-Bus - Raumtemperatur: 0.0 Â°C
FD 28 FF FF FF FF 00 00 DF
```
So if you want to work on a fix, fix these values first by setting them to valid parameters. The 0.2 and 0.1 values are a bit odd and may result from an issue in the EEPROM, but you can try setting them to sane values first and see what happens. Take note that in PPS mode, _all_ parameters that can be set are sent, continuously. The Set-button only stores them in BSB-LAN's memory/EEPROM. So just because you don't set a specific parameter doesn't mean that the parameter is not sent to the heater.

**Comment by fredlcore** [comment]
I have now fixed the TWW-Reduziertsollwert issue and also added debug warnings if some of the main temperature values are not above 0.

**Comment by malansk** [comment]
Thanks, I have tried it out but was not able to get it to send values because its complaining of invalid temperatur values.
I tried to set everything, but TWW Reduziertsollwert can't be set (SET Button missing).

**Comment by malansk** [comment]
![WebGUI-Setvalues](https://github.com/fredlcore/BSB-LAN/assets/25431062/52454d6c-6ac3-49dc-84fa-9424234ad665)

**Comment by malansk** [comment]
pps_values[PPS_RTS] != 0 might not work since this value is calculated by the heater and send to the QAA for info only, it can't be changed by QAA. As far as I know QAA tells the heater Anwesend Ein/Aus and from this the heater is giving RTS as output (depending on Komfortsollwert and Reduziertsollwert).

**Comment by fredlcore** [comment]
Good call, removed TWW-Reduziertsollwert in the check, so for now only room temperature comfort/reduced and TWW Nennsollwert are being tested against a value of 0.

**Comment by malansk** [comment]
It's still complaining. Maybe these values are after all 0 ?
How do they get populated - from EEPROM ? In the WEB-GUI the values are shown & they are persistend after restart.
Will test some more after dinner.

**Comment by fredlcore** [comment]
Sorry, I made a mistake there, of course the test must check for equal to zero, not the other way round. No need to download all again, just replace the line in 245 with this one:
`      if (pps_values[PPS_RTI] == 0 || pps_values[PPS_RTS] == 0 || pps_values[PPS_RTA] == 0 || pps_values[PPS_TWS] == 0) {`

**Comment by malansk** [comment]
The check is now working, no error after start (write mode disabled) but error as soon as I enter 21 degree as room temperature.

I did now uncomment all tx_msg from case1 until case26 but I still get the error.
However, now the heater controller switches from error to no error several times every few seconds.
Maybe not the content is the problem but the way how the pps_handling.h is called from the main program ?
Since I use the Arduino Due which is quite powerfull maybe the bus messages are transmitted to fast to too many at the same time overflowing some receiver buffer in the heater controller ?

**Comment by malansk** [comment]
[pps_handling.zip](https://github.com/fredlcore/BSB-LAN/files/15238546/pps_handling.zip)
[SerialMonitor_3.4.6_write_case0only_PPSBUSERROR.txt](https://github.com/fredlcore/BSB-LAN/files/15238551/SerialMonitor_3.4.6_write_case0only_PPSBUSERROR.txt)

**Comment by fredlcore** [comment]
You comment out the right lines as far as I can see quickly, but I see now that commenting out will still send the telegram, just one with a wrong checksum (which should just be ignored, not raise an error).
Try this:
Change this line:
` if (msg_cycle > 24 || (pps_values[PPS_QTP] == 0x52 && msg_cycle > 5)) {      // QAA50 sends fewer parameters`
to this
`    if (msg_cycle > 0 || (pps_values[PPS_QTP] == 0x52 && msg_cycle > 5)) {      // QAA50 sends fewer parameters`
That should only send FD 38 messages after a 17 request from the heater. The result should be that the connection is not interrupted, and the error might go away. It could still result in an error because now too few messages are coming, so you can increase the > 0 to > 1, > 2 etc. This will then send the telegrams from case 0, case 1 and case 2 (uncomment them before, of course). You can then change the contents of case 1 and 2 with other cases so that other telegrams are sent and see what happens.

If you really think that the code is the problem, just do a comarison between pps_handling.h between version 2.0.16 and the current version and adjust as you like and see if it changes anything (I'd bet it doesn't). 

Have you tried out using a different room unit model in the settings? Such as QAA50? That only sends 5 parameters, maybe this way it's easier to check? You still have to have valid values in the relevant parameters, though.

**Comment by malansk** [comment]
Changing the msg_cycle from 0 to 2 gets rid of the FF FF FF ... telegram as well as the "Verbindung unterbrochen 1 - ja" message due to too few messages.
But the heater still goes into error mode.
I have tried every possible option as QAA type but it does not change the behaviour.
I have searched the sourcecode of 2.0.16 for the include files but there are none, the program structure is completely different, all the functions are directly coded in the BSB-LAN.ino sketch.

**Comment by fredlcore** [comment]
Ok, the last thing I can think of then is to replace bsb.h and bsb.cpp from version 2.0.16 into the current version and see if that changes anything in case there are some weird timing issues involved. Otherwise, the only other way would be to install every version in-between to figure out which commit actually caused the error. By version I do not mean release versions, but when you go to "history" and see all the commits, you can click on the icon that gives you the state of the repository at the time of this commit. There you can download the repository as it was at that time, install it and see what happens. Take note that the BSB_LAN_config.h has changed sometimes over time, so you won't be able to use one for all the tests, but make sure that you set UseEEPROM to zero, so that you don't get a conflict with old EEPROM data.
Once you have figured out which commit actually caused your heater to go into error, I can have a closer look, but since I do not have a PPS heater myself and I just recently got confirmation from other PPS users that things are working fine with them, there is nothing else I can do unless you figure out which commit exactly causes the problem.

**Comment by fredlcore** [comment]
I thought again about whether it could be a timing issue, but I'm pretty sure that this can't be it, because the previous timing issues led to the heating system not accepting any commands that BSB-LAN sent because the answers from BSB-LAN didn't come quickly enough after a 0x17 (request to send) telegram from the heater.
In your case, the telegrams are accepted by the heater. The 0x38 telegram is accepted because the heater reacts with an answer that the connection is no longer interrupted. And you also mentioned that you can set other parameters. So the general communication on a hardware level is seemingly working fine. So I'm still pretty much convinced that it is one of the added telegrams that is causing the error. After reducing the telegrams sent to just the 0x38 telegram, have you checked the value in "Anlagentyp" in your heater again? It could be that it now stays at "15" (or whatever your correct value is) but that the error on the heater now comes from the fact that it is not getting enough parameter data that it would expect (e.g. it would assume the QAA to be faulty if it doesn't send the room temperature).

These are just some more thoughts that I had. If you think that all this is not the case and it's rather rooted in the code, you have no other way but to update your version step-by-step to figure out which version broke it. As some kind of "milestones" you could check the versions from the following dates first:
2 February 2021 (2.0.81)
24 September 2021
26 October 2021
24 March 2022
4 November 2022
12 December 2022
Figuring out where the error occurs in between these versions would be a first step, then you can move back/forth to identify which commit caused the issues...

**Comment by malansk** [comment]
Thank you very much for your support, in the meanwhile I have tried the following things.

1. Initialization of heater controller to actually start sending telegram on PPS bus
   I have realized that with PPS Read only mode, my heater is not sending out any data on the bus (not even 2.0.16 version) since 
  there is no "heartbeat" from QAA (I don't have a physical real QAA).
  So I always had to compile 2.0.16, run it in PPS write mode shortly to initialze the heater controller, compile new version (3.x).
  and do any trials.
 To reduce this pain, I have added hardcoded initilization to pps_handling.h:
========== 
if (bus->getBusType() == BUS_PPS) {
          byte tx_msg[] = {0xFD, 0x38, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x53, 0x7d}; //hardcode SEND(FD) QAAtype(38) QAA70(53) w. Checksum(7d)
          byte rx_msg[10] = { 0 };
          bus->Send(0, 0, rx_msg, tx_msg);
======
This successfully initailzes the heater controller even if PPS write is disabled and I get correct time and day shown in WEB-GUI als well as the temperature telegrams.

However, it also triggers the "Ungültiger Anlagentyp" error (Anlagentyp shown as "0" in heater controller).
Since the only thing send is the single "0xFD, 0x38, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x53, 0x7d" command, I think that the problem is comming from the low-level routines in bbs.cpp.
To verify this, I have exchanged the bbs.cpp and bbs.h with the old version from 2.0.16 but it's not working (compiler error) since there way a change from bool to int8 type for the Send () function.
I made a DIFF of the files bbs.h and bbs.cpp for your convenience and I tried to undo some of the changes in the new files but without getting rid of the error.
The differences as far as I understand:
*ESP32 support (should not be relevant since I use DUE) 
*new function BSB Disable Interface [took that out but no change of error]
*Support for additinal PPS telegram (F8, FB, FE) [took that out but no change of error]
*timing was changed to comply with BSB bus requiremens (min. delay  between 26ms .. 85ms changed to 63 .. 82ms [changed that back but no change of error]
*changed collission check with rx_pin while waiting  [changed that back but no change of error]
*omission of readbyte() after serial flush (read and discard the byte that was just sent to prevent it showing up in incoming buffer) [changed that back but no change of error]
*DC225 emulation (after sending 0x17, abort loop) [took that ou but no change of error]
*additional collision check on the bus (that's why I get the "collission on the bus, retrying"warning with the new version !) [took that out but no change of error]
*changed return value generation of the function [this part I was not shure how to undo - maybe you can help here ?]

**Comment by malansk** [comment]
[Windif_BSB.cpp_V2.0.16_vs_V3.4.6.pdf](https://github.com/fredlcore/BSB-LAN/files/15277021/Windif_BSB.cpp_V2.0.16_vs_V3.4.6.pdf)
[Windif_BSB.h_V2.0.16_vs_V3.4.6.pdf](https://github.com/fredlcore/BSB-LAN/files/15277022/Windif_BSB.h_V2.0.16_vs_V3.4.6.pdf)
[BSB.zip](https://github.com/fredlcore/BSB-LAN/files/15277052/BSB.zip)

**Comment by malansk** [comment]
Where is the link to that "History" page ?
On Github (https://github.com/fredlcore/BSB-LAN/releases) I only find 0.44 (working), 2.0 (which is in fact 2.1.8 from 31.07.2022 and already not working).

**Comment by fredlcore** [comment]
Here: https://github.com/fredlcore/BSB-LAN/commits/master/

**Comment by malansk** [comment]
[Windif_BSB.cpp_V2.0.16_vs_V2.1.8.pdf](https://github.com/fredlcore/BSB-LAN/files/15277132/Windif_BSB.cpp_V2.0.16_vs_V2.1.8.pdf)
[Windif_BSB.h_V2.0.16_vs_V2.1.8.pdf](https://github.com/fredlcore/BSB-LAN/files/15277133/Windif_BSB.h_V2.0.16_vs_V2.1.8.pdf)

**Comment by fredlcore** [comment]
You might again be too quick in jumping to conclusions, as I explained above that the error could result from too few telegrams coming from BSB-LAN which it would expect to be functioning properly. I don't have time into looking all the stuff you send me here, I told you what you have to do, even though it means that you cannot offload the work to me but have to do it yourself. Figure out when exactly the problem occurs and I'll have a look at it, but without having your exact controller, and knowing that your problem is only your problem and not any other PPS user's so far, you're the one who will have to do the work here...

**Comment by fredlcore** [comment]
Ok, since you get the "Collision detected" error sometimes, this was the only major change that also affected PPS, the rest of the changes didn't really affect PPS.
So, take the most recent version (that's what I ran a comparison against with the last version I have from April 2021), and look for:
```
  if (HwSerial == true) {
    serial->flush();
    unsigned long timeout = millis();
    while ((millis()-timeout < 50) && serial->available() == 0) {
      delay(1);
    }
    if (serial->available()) {      
      for (uint8_t i=0; i<=loop_len; i++) {
        char readdata = readByte();
        if (msg[i] != readdata) {
#if DEBUG_LL
          Serial.println(readdata, HEX);
#endif
          Serial.println("Collision on the bus, retrying...");
          delay(146);   // Wait the duration of 11 bits at 4800 bps times 32 (maximum telegram size) in ms (*1000) times 2 (because we can just keep waiting for an answer telegram) 
          while (serial->available()) {
            char c = readByte();
#if DEBUG_LL
            if (c < 16) Serial.print("0");
            Serial.print(c, HEX);
#endif
          }
#if DEBUG_LL
          Serial.println();
#endif
          goto retry;
        }
      }
    }
  } else {
```
and change it to
```
  if (HwSerial == true) {
    serial->flush();
#if defined(ESP32)
    unsigned long timeout = millis();
    while ((millis()-timeout < 50) && serial->available() == 0) {
      delay(1);
    }
    if (serial->available()) {      
      for (uint8_t i=0; i<=loop_len; i++) {
        char readdata = readByte();
        if (msg[i] != readdata) {
#if DEBUG_LL
          Serial.println(readdata, HEX);
#endif
          Serial.println("Collision on the bus, retrying...");
          delay(146);   // Wait the duration of 11 bits at 4800 bps times 32 (maximum telegram size) in ms (*1000) times 2 (because we can just keep waiting for an answer telegram) 
          while (serial->available()) {
            char c = readByte();
#if DEBUG_LL
            if (c < 16) Serial.print("0");
            Serial.print(c, HEX);
#endif
          }
#if DEBUG_LL
          Serial.println();
#endif
          goto retry;
        }
      }
    }
  } else {
#endif
```
and let me know if that changes anything.

**Comment by fredlcore** [comment]
It might be that the above solution still does not work or produce different problems. In that case, please send a log worth of 1-2 minutes of data. 
The problem is that everything BSB-LAN sends out on the bus comes back in again at the same time, but it first ends up in the UART buffer. Only when that one is full, we get notified. The problem is that our telegrams are smaller than the buffer (at least IIRC), so we might get notified "too late". That's why we have the 50ms delay at the beginning of the code in the  previous post which then just reads and discards the data we just sent to the bus, so that the buffer is clear afterwards.  It could be that the 50ms delay after sending a telegram is already "too long" for your PPS system. Meaning, after we have sent our data, the heater has already started sending its next telegram within the 50ms pause for BSB-LAN. So when we then read and comare the data to what we have just sent, it doesn't match and therefore you get the "Collision" warning, which leads to BSB-LAN sending the previous telegram again - which then does not match the expectation of the heater.
So this would be a somewhat possible explanation of why there is a problem now compared to the past where we didn't wait ater a send. But if the problem persists after removing this clearing of the UART input buffer, then it will become difficult: Because we don't know how long exactly we would have to wait: It has to be long enough to receive the just sent telegram, but short enough not to miss the next telegram from the heater.
The only hope could come from the fact that we know the length of a PPS telegram sent by BSB-LAN/QAA, it's always 9 bytes. 9 bytes with 10 bits each (8O1) at 4800bps take 2.08 ms per Byte, so that would be 18.72ms. So you could change the 50 to 20 and see if that makes a difference. 
But now that I think about it, we may not even have to wait that long: After sending out the first byte, it will arrive in the receiving buffer of the UART while we send out the second byte, and the second byte will reach us while we send the third byte etc. So it would be possible that we only have to wait for 2,08 (let's make it 3) milliseconds. So as a next step (if 20ms doesn't do the trick), you could set it from 50 to 3.
In parallel, I'll try this with my LPB system at home and see if that does the trick as well.

**Comment by malansk** [comment]
I was working in parallel with the history of the commit in 2022 and found that this commit from 23.03.2022 breaks it for my system:
https://github.com/fredlcore/BSB-LAN/commit/03b8aff9aab004cacf1c7548841405f01464426f

I have done the following changes to the most recent master version 3.4.6:
===============
pps_handling.h
/* //Martin Start Uncomment
	if ((msg[0] & 0x0F) == 0x0E) {   // Heater requests information from the QAA (i.e. BSB-LAN) with telegram type 0x1E [original version from 3.4.6]
	//*/ //Martin End Uncomment
	if (msg[0] == 0x1E) {   // Heater requests information from the QAA (i.e. BSB-LAN) with telegram type 0x1E [old version before commit from 23.03.2022]  

==============
bsb.cpp
// ... until SOF detected (= 0xDC, 0xDE bei BSB bzw. 0x78 bei LPB)
	//Martin Debug Start 
	/*
	if ((bus_type == BUS_BSB && (read == 0xDC || read == 0xDE)) || (bus_type == BUS_LPB && read == 0x78) || (bus_type == BUS_PPS && ((read & 0x0F) == 0x07 || (read & 0x0F) == 0x0D || (read & 0x0F) == 0x0E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE))) {    // PPS telegram types 0x17, 0x1D, 0x1E, 0xF8, 0xFB, 0x$FD and 0xFE
    */ //Martin Debug End
	if ((bus_type == BUS_BSB && (read == 0xDC || read == 0xDE)) || (bus_type == BUS_LPB && read == 0x78) || (bus_type == BUS_PPS && (read == 0x17 || read == 0x1D || read == 0x1E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE))) {    // PPS telegram types 0x17, 0x1D, 0x1E, 0xF8, 0xFB, 0x$FD and 0xFE
      // Restore otherwise dropped SOF indicator
      msg[i++] = read;
      //Martin Debug Start
	  /*
	  if (bus_type == BUS_PPS && (read & 0x0F) == 0x07) {   // PPS RTS telegram (0x17)? [original version from 3.4.6]
	  */ //Martin Debug End
	  if (bus_type == BUS_PPS && read == 0x17) {   // PPS RTS telegram (0x17)? [old version before commit from 23.03.2022]

With these changes I no longer get the Error Message and I am able to write setting from the Web-GUI to the controller :-)

**Comment by malansk** [comment]
I have checked now how the heater controller reacts to temperature value settings of 0°C or negative °C for Raumtemperatur Istwert, Raumtemperatur Nennsolweert, Raumtemperatur Reduziertsollwert, Trinkwasser Sollwert, Trinkwasser Reduziertsollwert. It's not problem to send 0 or even negative values, this does not generate an error. The heater defaults in such cases to its own internal limit (0° / 4° / 8° / ...)
So it would be probably safe to take out the check for illegal temperature values since it might confuse some people why no write enabled (if they don't check the SerialMonitor).

**Comment by malansk** [comment]
I have checked the different timings (50 / 20 / 10 /3ms) on the original (without my patches to undo commit from 23.03.20222) 3.4.6 version but it does not make the heater error (or the SerialMonitor message) go away.

I have then also check on my patched version. 50 / 20 / 10 does not make a difference - heater shows no error, i can set values, SerMon Message (Collission...) is still coming one time after startup.
3ms is giving me the heater error (ungültiger Anlagentyp) - probably its flooding the bus due to too small delay and this errormessage from the heater is just a general "PPS-BUS" error.

**Comment by malansk** [comment]
Some more observations with my patched version:
*Setting of DAY of the week from GUI to heater is working
*Setting of TIME from GUI to heater results in change of DAY and time (get set to 00:00)
 Since this was read-only in 2.0.16 I can't tell if this behaviour is due to my patch or if it is a separate issue of my heater or of BSB-Lan. But I don't need this feature so it's no problem for me if this is not working.
*Außentemperatur gemischt and Vorlauftemperatur Sollwert have previously always shown 0°C, now the show 292°C and 220°C. Maybe the commit from 23.03.2022 should have only been done on PPS_MCBA device family and not on the standard PPS ? Vorlauftemperatur Sollwert is only available on PPS_MCBA as far as I know. Außentemperatur gemischt is available on my heater controller but not shure if this gets transmitted over the bus (probably not).

**Comment by fredlcore** [comment]
Ok, thanks for the testing. It makes sense to me now, but I don't know how to solve this matter:
The code you reverted in bsb.cpp handles the detection of the beginning of a telegram. When BSB-LAN starts to listen to the data stream on the bus and it detects some kind of data, it doesn't know if this data is coming from the middle of a telegram or not. That's why it has to look out for some "markers". With BSB and LPB, that's relatively easy, they have one specific marker (DC and 78 respectively) that I can look out for. With PPS, it's a bit more difficult, because it has various markers: 17, 1D, 1E, F8, FB and FD. To make matters more complicated, some PPS heating devices increase the high nibble (the "left" part of the byte) with each iteration by one or two. So, instead of sending 17 only, they would send 17 the first time, 27 the next time, 37 thereafter etc. The same goes for the 1D and 1E type of telegrams.

For that reason, I had to change the code in 2022 to not look at the complete byte, but just at the lower nibble to test it for 7, D or E. That is what `(bus_type == BUS_PPS && ((read & 0x0F) == 0x07 || (read & 0x0F) == 0x0D || (read & 0x0F) == 0x0E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE)` does.
The problem is that since we don't know whether we are in the middle of a telegram or not, this test will also match if, let's say, a 27 is part of a telegram that the heater just sends out for information and which we just happen to intercept. Since PPS is very time-critical, BSB-LAN thinks it's a "request to send" telegram (that is what the type 17 telegrams are) and immediately starts to send the requested response - which in that case is of course wrong because it will lead to sending an unwanted telegram, potentially even during a time when the heater is still sending its own telegram (potentially resulting in a collision on the bus).

The reason yours is working now is that the chances of accidentially hitting a 0x17 somewhere in a regular telegram are 15 times less than hitting a byte that has a 7 in the lower nibble (which could also be 0x07, 0x27, 0x37 etc. all the way to 0xF7).
With other telegram types that is not a problem because there will always be a checksum at the end, so once the telegram is received completely, BSB-LAN knows whether it was an accidental hit (then the checksum is wrong) and then discards the telegram. That is, however, not the case with the "request to send" telegram 0x17: Other than all other PPS telegrams (which consist of nine bytes), this only consists of one single byte: 0x17 (or on some machines also 0x27 etc.).
And since PPS requires the QAA to respond more or less immediately (even printing some debug messages on the serial port take too long and will result in a timeout), BSB-LAN has to send the answer telegram immediately which then can result in the problems you are facing.

However, for the same reasons, I don't think that the changes in pps_handling.h are necessary. Here, we already have a complete and validated telegram at hand, and we don't test for it on any position, but only at the very beginning. Also, whether we test for 0x1E or generally for a lower nibble of E shouldn't make a difference because I know of no telegram type on PPS that has a lower nibble of E (there are also none in your logs). So could you please revert the changes in pps_handling.h back to the state of the current version and try again if your heater still doesn't show any error?

Since I have no way to detect if a PPS system only sends 0x17 or also any other byte with a lower nibble of 7, I cannot simply revert back to your changes in bsb.cpp for the general public. And if my assumptions are true, you might still run into the same problem, just somewhat less likely. The only reason I can think of now is not only to test for 0x17/lower nibble 7, but also to wait the amount of time it takes for the next byte to come in. If there is no more data coming in, the 0x17/lower nibble 7 must be the "request to send" telegram. Then the routine can continue as it is now. If there is another byte within that time, the 0x17/lower nibble 7 is part of a larger telegram, so we must wait for its end first.

I'll see if this can be done or if even such a short delay is already too much for the tight PPS timing. But as I said, if your could try and test reverting the changes in pps_handling.h, that would be great.

**Comment by fredlcore** [comment]
I have now made some changes in bsb.cpp. Could you please do the tests in the following order:
1. Revert the changes you made in your pps_handling.h version and check if these were really necessary (as discussed in my previous post)
2. If these are not necessary, install the most recent version from master repo. If they are still necessary, make these changes again in the most recent pps_handling.h as well
3. With the current version, the errors on your heater should not appear, but it might be that you are getting "connection interrupted: yes" as a result. In that case, change this line:
`       while (millis()-wait_timer <= 4) {    // wait up to 4ms for a new byte to be reported by HardwareSerial`
and reduce the value `4` down to 3, 2, 1 and 0 until the "connection interrupted: yes" messages are going away and you have a stable connection.
4. If this works, let me know which threshold value works reliably, so I can use that for the future.

Thanks

**Comment by malansk** [comment]
1. The changes in pps_handling.h are not necessary.
2. I have installed the latest version without any patches from my side
3. The heater will go into error mode after a few seconds, no matter which delay i try (25, 4, 3, 2, 1, 0). It is always showing "Verbindung unterbrochen 0 - NEIN".
 I am surprised that it goes into error mode since with RTI defaulting to 0 it should not try to write to the bus ?
 I took out the checks for 0 and also tried to set temperature but it still gets to error mode.

**Comment by malansk** [comment]
[SerialMonitor_PPS_Bugfix_11.05.2024.txt](https://github.com/fredlcore/BSB-LAN/files/15282052/SerialMonitor_PPS_Bugfix_11.05.2024.txt)

**Comment by malansk** [comment]
Behaviour with this version:
*Start in Read-Only mode
*After Setting of RTI it will run for approx. 5..10s then it will throw the Error.
*Tried 1,2,3,4, 10, 20ms value in BBS.cpp

What's a good way to capture longer SerialMon Output duration ?
Arduino IDE dos not provide an option to copy/past a.f.a.i.k. so I use Putty (Copy to Clipboard) but this is limited to approx. 60s then it will start to overwrite the beginning of the log.
I will for now send two part log, first part is startup, setting RTI, heater error. Second part shows Connection Status. It will remain showing "Verbindung verloren 0 - Nein" even after the heater error.

**Comment by malansk** [comment]
[SerialMonitor_PPS_Bugfix_11.05.2024_14_02.txt](https://github.com/fredlcore/BSB-LAN/files/15282718/SerialMonitor_PPS_Bugfix_11.05.2024_14_02.txt)
[SerialMonitor_PPS_Bugfix_11.05.2024_14_02_Part2.txt](https://github.com/fredlcore/BSB-LAN/files/15282719/SerialMonitor_PPS_Bugfix_11.05.2024_14_02_Part2.txt)

**Comment by fredlcore** [comment]
You should be able to set the screen buffer size in putty's settings somewhere, I guess...

**Comment by fredlcore** [comment]
Ok, can you check again if the error goes away when you revert to the changes you made above (i.e. checking for 0x17 only)?

**Comment by fredlcore** [comment]
Also, please make these changes in bsb.cpp:
```
          if (serial->available() > 0) {
            break;
          }
```
to
```
          if (serial->available() > 0) {
Serial.println("X");
            break;
          }
```


```
        if (serial->available() == 0) {
          return true;  // No more bytes incoming? PPS-Bus request byte 0x17 just contains one byte, so return
        }
```
to 
```
        if (serial->available() == 0) {
Serial.println("Y");
          return true;  // No more bytes incoming? PPS-Bus request byte 0x17 just contains one byte, so return
        }
```

```
      // Delay for more data
      if (HwSerial == true && serial->available() == 0) {
        delay(4);   // I wonder why HardwareSerial needs longer than SoftwareSerial until a character is ready to be processed. Also, why 3ms are fine for the Mega, but at least 4ms are necessary on the Due
      }
```
to
```
Serial.println("Z");
      // Delay for more data
      if (HwSerial == true && serial->available() == 0) {
        delay(4);   // I wonder why HardwareSerial needs longer than SoftwareSerial until a character is ready to be processed. Also, why 3ms are fine for the Mega, but at least 4ms are necessary on the Due
      }
```

Also please remove the temperature checks if you are sure that these have nothing to do with the error. Then please send me a log from boot until "Verbindung unterbrochen 0 - Nein".

**Comment by malansk** [comment]
I did my patch, above debug changes and removed check for 0 values. There is no error shown and I can set values.
I did set RTI to 21.5 and Betriebsart from Manual to Auto.

Still need to figure out Putty buffer size...

**Comment by malansk** [comment]
[SerialMonitor_PPS_Bugfix_11.05.2024_16_15.txt](https://github.com/fredlcore/BSB-LAN/files/15283123/SerialMonitor_PPS_Bugfix_11.05.2024_16_15.txt)

**Comment by fredlcore** [comment]
Ok, thanks, can you now do one more log run without your changes, but with the X/Y/Z debug messages in place?

**Comment by fredlcore** [comment]
And it would help (also for log size) to disable MQTT logging during the tests...

**Comment by fredlcore** [comment]
By the way: The "collision on the bus" comes as a result from the parameter logging. Since the heater is continuously firing data and requests on the bus (about two telegrams per second), these do not get served immediately and thus pile up in the serial receive buffer. So once logging is done, these telegrams are processed by BSB-LAN which not only sends out answers to telegrams that the heater no longer knows it has requested, but the answers sent out by BSB-LAN do not show up on the bus (because they would only come up later in the buffer, after all the previously buffered telegrams are processed. This is interpreted by BSB-LAN as a collision on the bus even though technically, there need not have been one (but it could) sent and received telegrams do not match. This leads to the whole receive buffer being discarded.
It doesn't hurt and is probably also the most sensible way to deal with this, but just to explain where this comes from.

**Comment by fredlcore** [comment]
Once you have generated the log file with the original conditions (instead of just 0x17), could you please try to figure out which condition exactly results in the error?
So above you had made these changes:
```
	if ((bus_type == BUS_BSB && (read == 0xDC || read == 0xDE)) || (bus_type == BUS_LPB && read == 0x78) || (bus_type == BUS_PPS && ((read & 0x0F) == 0x07 || (read & 0x0F) == 0x0D || (read & 0x0F) == 0x0E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE))) {    // PPS telegram types 0x17, 0x1D, 0x1E, 0xF8, 0xFB, 0x$FD and 0xFE
```
to
```
	if ((bus_type == BUS_BSB && (read == 0xDC || read == 0xDE)) || (bus_type == BUS_LPB && read == 0x78) || (bus_type == BUS_PPS && (read == 0x17 || read == 0x1D || read == 0x1E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE))) {    // PPS telegram types 0x17, 0x1D, 0x1E, 0xF8, 0xFB, 0x$FD and 0xFE
```
So if you take the first one and then step by step change from e.g. `(read & 0x0F) == 0x07` to `read == 0x17` and so on and so forth to figure out which condition is the one we need to investigate further?

Also, Ffom my understanding, this change should not be necessary:
```
	  if (bus_type == BUS_PPS && (read & 0x0F) == 0x07) {   // PPS RTS telegram (0x17)? [original version from 3.4.6]
```
to
```
	  if (bus_type == BUS_PPS && read == 0x17) {   // PPS RTS telegram (0x17)? [old version before commit from 23.03.2022]
```
because based on your condition above, it filters for 0x17 anyway, right?
Anyway, let's see which condition is the troublemaker here. I hope it's just one and not several one of them...

**Comment by fredlcore** [comment]
Maybe before you do any of the above, I have another idea which might be easier to check:
Take the most recent version from GitHub and leave bsb.cpp as it is. Then only make this change in pps_handling.h, from 
```
if ((msg[0] & 0x0F) == 0x07 && msg[0] != 0xB7 && pps_write == 1) {
```
to
```
if (msg[0] == 0x17 && pps_write == 1) {
```
This is the part of the code where the heater is reacting to the 0x17 request to send telegram. So even if the checks in bsb.cpp are too loose, nothing should happen nevertheless, because this is the place where BSB-LAN would send its own stuff. 
If this already fixes things and makes your error go away, we can concentrate on this part and not on the low-level bus parts. In that case, revert the condition back to the first one and try again, as I have included some sanity-checks further down below. Let's see how that goes...

**Comment by fredlcore** [comment]
If making the changes in the previous post do not change anything (meaning the changes _have_ to be in bsb.cpp to prevent the error), then please continue with the tests from my posts before.

**Comment by malansk** [comment]
The most recent version hangs (Web-GUI no longer reachable, no output in SerialMonitor) after trying to set RTI via WebGUI

======
1D 4F FF FF FF FF FF 01 98
PPS RTS 
17 00 00 00 00 00 00 00 00
Not sending values to heater due to invalid temperature values in BSB-LAN.
PPS REQ 15046.0 PPS-Bus - QAA Modell: 
4E 38 FF FF FF FF FF FF 80
GET /S15008=21.0 HTTP/1.1

set ProgNr 15008 = 21.0
15008.0 PPS-Bus - Raumtemperatur: 21.0 °C
00 28 00 00 00 00 05 40 00
#15008: 21.0 °C
Publishing to topic: BSB-LAN/15008
Payload: 21.0
Successfully published...
PPS RTS 
47 00 00 00 00 00 00 00 00
PPS INF 15045.0 PPS-Bus - Verbindung unterbrochen: 1 - Ja
1D 4F FF FF FF FF FF 01 98

**Comment by fredlcore** [comment]
Can you turn off MQTT logging please? And if you say that it "hangs", does it mean the output on the serial monitor also stops?

**Comment by malansk** [comment]
After turning off MQTT logging, it no longer hangs.
With the unmodified version, there is the heater error.
If I patch to:
if (msg[0] == 0x17 && pps_write == 1) {
There is no heater error but the RTI value is not transmitted to the controller (even though write mode gets enabled) and there are a lot of PPS REQ for QAA type (even thou QAA type is announced by BSB-LAN).

**Comment by malansk** [comment]
[SerialMonitor_13.05.2024.txt](https://github.com/fredlcore/BSB-LAN/files/15290959/SerialMonitor_13.05.2024.txt)

**Comment by fredlcore** [comment]
That's exactly what I expected. Because your heater not only uses the 0x17 for RTS (request to send), but also 0x47, the modification leads to the fact that BSB-LAN no longer responds to 0x47 telegrams. The log stops relatively early, but it should only be a matter of time until it reports that the connection is interrupted.

**Comment by fredlcore** [comment]
What I don't understand is how this worked for you before. In 2.0.16, we didn't know about any other telegram than 0x17 that the heater would use to indicate RTS. So with this modification, it should be as back then. So if you say it's now no longer reacting to RTI, I'm surprised that it did so in 2.0.16.

**Comment by fredlcore** [comment]
Can you send me another log of the unmodified, most recent version (I removed the sanity checks now) and if possible try to identify the moment that the heater goes into error and the telegram that is sent around that time?
And also send me another log with version 2.0.16, because the one above was made with monitor setting.
We really need to identify what exactly makes the difference here...

**Comment by malansk** [comment]
In parallel I have made some tests to the version from Saturday.
I did only modification to this line:
if ((bus_type == BUS_BSB && (read == 0xDC || read == 0xDE)) || (bus_type == BUS_LPB && read == 0x78) || (bus_type == BUS_PPS && ((read & 0x0F) == 0x07 || (read & 0x0F) == 0x0D || (read & 0x0F) == 0x0E || read == 0xF8  || read == 0xFB || read == 0xFD || read == 0xFE))) {    // PPS telegram types 0x17, 0x1D, 0x1E, 0xF8, 0xFB, 0x$FD and 0xFE

The 0x07 statement throws the error so I changed this to read == 0x17.
After this modification, no more error but repeated request INF QAA type.
If I then remove 0x0D the repeated requests INF QAA type stop and it is working.

**Comment by malansk** [comment]
With the most recent version (sanity check removed) the log shows the last 1 or 2 seconds after the heater went into error.

**Comment by malansk** [comment]
[SerialMonitor_13.05.2024.10.53_unmodified.txt](https://github.com/fredlcore/BSB-LAN/files/15291929/SerialMonitor_13.05.2024.10.53_unmodified.txt)

**Comment by malansk** [comment]
[SerialMonitor_2.0.16_write_MonitorOff.txt](https://github.com/fredlcore/BSB-LAN/files/15292010/SerialMonitor_2.0.16_write_MonitorOff.txt)

**Comment by fredlcore** [comment]
I'm not sure if I understand you right. I need the logs from current version from boot of ESP32 until the heater goes into error (and a bit more if possible). At best, you could mark the moment the heater goes into error, so I know roughly where to take a look.
As for changes in bsb.cpp: Did you just remove the `(read & 0x0F) == 0x0D` or did you change it to 0x1D? Because if you just removed it, it means that the INF telegrams are going to be ignored. You don't want that because they also contain data for several parameters, for example whether the connection is interrupted or not. If should be visible in the log file that this information is then missing.

**Comment by fredlcore** [comment]
Generally, the amount of log data is too little to make a thorough analysis. On PPS, some telegrams are only sent every couple of minutes. This is what I found for putty on Google:
> By default, PuTTY sets the scrollback buffer to 200 lines, which may not be enough for some tasks or situations. To increase or decrease the scrollback buffer, open PuTTY and go to the Window category on the left pane. Then, enter the number of lines that you want in the Lines of scrollback box.

So, set it to 2000 or 10000, that should give us more room to breathe...

**Comment by fredlcore** [comment]
I think I got it now. In my old e-mails, I found another user with a RVA53 who also reported an (unspecified) error when BSB-LAN was in room unit mode. And a user with a LGM11 which sends 0xB7 (in addition to 0x17), to which we also found out that we must not react to. It might be that the 0x17 are addressed to the QAA, whereas the 0x47 and 0xB7 are addressed to other devices on the bus (which, on a _P_oint-to-_P_oint-_S_ystem doesn't really make sense). There was, however, one user where the changing high nibble had to be observed, so I have to find that one again and see if I can generate a rule from there, but it seems that your system is similar in the way that the 0x47 has to be ignored because the heater thinks it comes from some other kind of device and is thus confused if this other type of device identifies as a QAA70...

**Comment by malansk** [comment]
The Putty settings I had already changed, maybe Notepad has here a limitation, I now use wordpad to save the log.
Current version without modification -> error after a few seconds after enabling write mode
Current version with modification from 0x07 to 0x17 -> no error but RTI not taken, repeated requests for QAA type
Current version with modifcation from 0x07 to 0x17 and removal of 0x0E -> working as it should

**Comment by malansk** [comment]
[SerialMonitor_change_to17_removal_of_0E_5minlog.txt](https://github.com/fredlcore/BSB-LAN/files/15293565/SerialMonitor_change_to17_removal_of_0E_5minlog.txt)

**Comment by fredlcore** [comment]
Ok, I added a new QAA type RVD130 since this was the only heater that changed the high nibble for the magic byte in every telegram. Can you please download the most recent version and see how this behaves? I hope that should do it now...

**Comment by malansk** [comment]
![MQTT-hang](https://github.com/fredlcore/BSB-LAN/assets/25431062/93ad7db5-cfd5-41d8-a632-7f53a9f9b0aa)
[SerialMonitor_NewQAAType_MQTT-hangup.txt](https://github.com/fredlcore/BSB-LAN/files/15295953/SerialMonitor_NewQAAType_MQTT-hangup.txt)

**Comment by fredlcore** [comment]
Good to know. But the logs say that the data has successfully been published via MQTT. Can you reduce the number of logged parameters? And why is there a lot of access to /Exxxxx URLs?

**Comment by malansk** [comment]
[MQTT-hang_OneWireeEnabled_DHTDisabled.txt](https://github.com/fredlcore/BSB-LAN/files/15296368/MQTT-hang_OneWireeEnabled_DHTDisabled.txt)
![mqtt-hang_DHT-disabled_OneWireEnabled](https://github.com/fredlcore/BSB-LAN/assets/25431062/c6b1614a-72e1-4e5d-9a55-d576f6208e05)
[MQTT-hang_DHT-disabled_OneWire-Enabled.pdf](https://github.com/fredlcore/BSB-LAN/files/15296369/MQTT-hang_DHT-disabled_OneWire-Enabled.pdf)

**Comment by fredlcore** [comment]
So what does this mean now?

**Comment by malansk** [comment]
I just realized that there must have been some old data in the EEPROM, I now set "use EEPROM" to 0.
This gets rid of PPS value publishing on MQTT but still publishes the 20000 data's.
Will test some more later

**Comment by malansk** [comment]
So I have started with a fresh default config.h (with #define RESET but no changes to logging settings) and with a /NE command.
Then I have put in MQTT credentials end enabled logging via WEB-GUI.
-> BSB-LAN publishes Status=Online and 15002, 15006, 15004, 15064 values Once and then hangs.
I cant's figure out where it gets the configuration to publish 15002, 15006, 15004 and 15064 - this was my original configuration for 2.0.16 version but I have done /NE (w. #define RESET and it really resets some values) and used default config.h  - maybe the EEPROM does not get fully erased ?

**Comment by malansk** [comment]
[MQTT-hang_EEPROM-erased_defaultConfig.h.txt](https://github.com/fredlcore/BSB-LAN/files/15297417/MQTT-hang_EEPROM-erased_defaultConfig.h.txt)
[BSB_LAN_config.zip](https://github.com/fredlcore/BSB-LAN/files/15297427/BSB_LAN_config.zip)
[MQTT-hang_EEPROM-erased_defaultConfig.h.pdf](https://github.com/fredlcore/BSB-LAN/files/15297446/MQTT-hang_EEPROM-erased_defaultConfig.h.pdf)

**Comment by fredlcore** [comment]
Since a few weeks, BSB-LAN publishes all values that are queried via the web-interface to MQTT (if MQTT is enabled). I saw in your logs that you have a number of URL-based queries, some also with /E... to query the ENUM values which I don't know why you do that. But whenever these URLs are called to read a parameter, it's also published to MQTT:
```
GET /E15002 HTTP/1.1
GET /20301 HTTP/1.1
GET /E20301 HTTP/1.1
GET /20303 HTTP/1.1
GET /E20303 HTTP/1.1
GET /15006 HTTP/1.1
```

And what does "hang" mean? Your log shows a number of MQTT publishing of which 15064 is the last one at some point, but the log continues after that.

**Comment by malansk** [comment]
Now I have put in 15000,15001,15002,15003,15004,15005,15006,15007,15008 via WEB-Gui as logging Parameter & saved that config via WEB-Gui.
-> Now it publishes 15000,15001,15002,15003,15004,15005,15006,15007,15008  - but also 15064.
So for me it looks like indeed the /NE did not fully erase the EEPROM, only with overwriting it from Web-Gui it took the new settings (but has somehow continued to use the old 15064 setting).
Furtunately, now it does not hang (SerMonitor continues to update and Web-GUI is reachable.).
Before i put in the new logging parameters vie Web_Gui it hang (SerMon stopped upadting, We-Gui no longer reachable), had to restart Due).

**Comment by fredlcore** [comment]
It would really help if you could just answer my questions first before making your own guesses. Where do the URL calls come from? This one here is from your last log:
```
GET /15064 HTTP/1.1
15064.0 PPS-Bus - Heizperiode 2 Mittwoch Start: 00:00-00:00, 00:00-00:00, 00:00-00:00
00 62 00 00 00 00 00 00 00
#15064: 00:00
Publishing to topic: BSB-LAN/15064
Payload: 00:00
Successfully published...
GET /E15064 HTTP/1.1
```
So something is calling /15064 as well as /E15064 regularly. It has nothing to do with the EEPROM, it's probably your home automation system that is making these calls. If not, then you have a problem with an intruder because BSB-LAN doesn't call itself...

**Comment by malansk** [comment]
Yes, I have just now realized in parallel to your comment that this is from CuxD running on CCU3 homeautomation system. I had disabled writing to BSB-Lan some days ago (due to changed PPS addresses from 2.016 to 3.4.6) but forgot to disable reading via GET command from BSB-Lan.

**Comment by fredlcore** [comment]
Ok, so if everything is running fine now, can you please download the most recent version one more time and use your system "normal" (with setting verifiable values on the heater, with MQTT publishing etc.) and let me know if everything runs hopefully fine? Thank you.

**Comment by malansk** [comment]
So now it is running very good with PPS write mode fixed and MQTT working as good as ever ;-)
Thank you very much for your time and effort for this update !
I will now update my home automation system to the new PPS Program Numbers and keep the 3.4.6 version :-)
Would be nice if you could publish an "official" new version on Github in the near future so I would use that "official" version and not the current testbuild from the Master repo.

**Comment by fredlcore** [comment]
Master is not a "testbuild", it is always the most recent version. "Releases" are just milestones that are published when new (sometimes breaking) changes are introduced, so that users can know which version is the last one for their board type etc. So it depends on the feature change, sometimes there is just one per year, sometimes more.

**Comment by malansk** [comment]
Ok, then I will use the current Master.

**Comment by fredlcore** [comment]
@malansk We are about to release version 3.4 and had to make some changes in the data transfer for BSB/LPB. Could you please download the most recent master repo version and check if BSB-LAN still works in active/room unit mode (no permanent "Connection interrupted" etc.)?
Thank you!
